numba/misc/coverage_support.py

```diff
@@ -79,22 +61,138 @@ class NotifyCompilerCoverage(NotifyLocBase):
+        but instead receive fake trace event for each line the compiler have
+        visited.
+
+        See coverage.PyTracer
+        """
+
+        data: coverage.types.TTraceData
+        trace_arcs: bool
+        should_trace: coverage.types.TShouldTraceFn
+        should_trace_cache: Mapping[
+            str, coverage.types.TFileDisposition | None
+        ]
+        should_start_context: coverage.types.TShouldStartContextFn | None
+        switch_context: Callable[[str | None], None] | None
+        lock_data: Callable[[], None]
+        unlock_data: Callable[[], None]
+        warn: coverage.types.TWarnFn
+        packed_arcs: bool
+
+        _activity: bool = field(default=False)
+
+        def start(self) -> coverage.types.TTraceFn | None:
+            """Start this tracer, return a trace function if based on
+            sys.settrace."""
+            return None
+
+        def stop(self) -> None:
+            """Stop this tracer."""
+            return None
+
+        def activity(self) -> bool:
+            """Has there been any activity?"""
+            return self._activity
+
+        def reset_activity(self) -> None:
+            """Reset the activity() flag."""
+            self._activity = False
+
+        def get_stats(self) -> dict[str, int] | None:
+            """Return a dictionary of statistics, or None."""
+            return None
+
+        def trace(self, loc: ir.Loc) -> None:
+            """Insert coverage data given source location.
+            """
+            # Check whether the file should be traced
+            disp = self.should_trace_cache.get(loc.filename)
+            if disp is None:
+                disp = self.should_trace(loc.filename, None)
+                self.should_trace_cache[loc.filename] = disp
+            if not disp.trace:
+                # Bail if not tracing the file
+                return
+            # Insert trace data
+            tracename = disp.source_filename
+            self.lock_data()
+            cur_file_data = self.data.setdefault(tracename, set())
+            if self.trace_arcs:
+                if self.packed_arcs:
+                    cur_file_data.add(_pack_arcs(loc.line, loc.line))
+                else:
+                    cur_file_data.add((loc.line, loc.line))
+            else:
+                cur_file_data.add(loc.line)
+            self.unlock_data()
+            # Mark activity for this tracer
+            self._activity = True
+
+    def _pack_arcs(l1: int, l2: int) -> int:
+        """Pack arcs into a single integer for compatibility with .packed_arcs
+        option.
+
+        See
+        https://github.com/nedbat/coveragepy/blob/e7c05fe91ee36c0c94e144bb88d25db4fc3d02fd/coverage/ctracer/tracer.c#L171
+        """
+        packed = 0
+        if l1 < 0:
+            packed |= 1 << 40
+            l1 = -l1
+        if l2 < 0:
+            packed |= 1 << 41
+            l2 = -l2
+        packed |= (l2 << 20) | l1
```

Comment:
```suggestion
        packed |= (l2 << 20) + l1
```
original code is `+` (see https://github.com/nedbat/coveragepy/blob/e7c05fe91ee36c0c94e144bb88d25db4fc3d02fd/coverage/ctracer/tracer.c#L192)